CAM reanalysis bias correction for CLM-DART

braczka opened the pull request:
## Description:
Scripting has been added that corrects the meteorological bias in the CAM reanalysis forcing when performing a site-level assimilation with CLM-DART.   This bias is rooted in the spatial mismatch between a CAM grid cell and the spatial coverage of a land site (typically 1x1 km2).   The scripting includes nco commands that will cut out the CAM grid cell collocated with the land site of interest.  Subsequent scripts perform the bias correction, while still maintaining necessary ensemble spread to generate model error covariance in CLM.   


### Fixes issue
This PR addresses issue #446  CAM reanalysis (site-level) bias correction tool

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

### Documentation changes needed?
- [X] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly.

I have provided some brief documentation as comments within the scripts themselves, and also a README  file as part of the PR.  Eventually I will add web-based documentation, but I still want to complete more testing before fully advertising the code.

### Tests
Please describe any tests you ran to verify your changes.

These scripts have undergone significant diagnostic checks (some diagnostic figures are included in the PR), and also have been successfully run with several site level ensemble free runs.  For example see CLM-DART case run:  /glade/work/bmraczka/cases/cesm2.2.0/clm5_CAM4_BGC_analysis_adjusted_e80 or case run:  /glade/work/bmraczka/cases/cesm2.2.0/clm5_NR1_BGC_analysis_adjusted_e80.    The bias correction code still needs additional testing with an assimilation run, but the code is ready to receive software review for improvements/streamlining.

## Checklist for merging

- [x] Updated changelog entry
- [ ] Documentation updated
- [x] Update conf.py

## Checklist for release
- [ ] Merge into main
- [ ] Create release from the main branch with appropriate tag
- [ ] Delete feature-branch

## Testing Datasets

- [x] Dataset needed for testing available upon request
- [ ] Dataset download instructions included
- [ ] No dataset needed

The scripts include paths for an example run for flux tower site US-NR1.  

For example:

```
% Input data files
 path_CAM_forcing = '/glade/work/bmraczka/CAM6_NR1/';
 path_towermet_forcing = '/glade/work/bmraczka/SIFMIP2/tower_met_forcing/';
```

kdraeder reviewed (commented):
`CAM6_bias_correct.m `
(similar comments likely apply to CAM4_bias_correct.m)

Reduce code by filling enstr using a loop of formatted writes?
   num2str?  printf?

Trying to make code less offensive to a wider audience;
   change 'master' to something else?  main, root, reference, ...?

(I may just need a Matlab review) Why does X_master need to be defined, but empty,
then used in a redefinition [X_master;X_dummy]?
Then X_dummy are cleared, so why aren't X_master filled directly by the ncreads?

253: Why is swvdr in there twice?  2nd (4th term) should be swvdf?
257-265; the denominator in all 4 looks identical to me,
         and it may be the same sum as in the if-test,
         so calculate it once into a new variable, and use that 5 times?
323-326; why concatenate 2 lines into 1 long line?
  later; more of the same

There's a lot of redundancy in the plot generation.
Did you already consider defining a function(s) to use instead?
If not, I'm willing to fetch the code and figure out how complicated it would be.
If I do that, I could use some advice for the best way to get the code.

hkershaw-brown commented on models/clm/CAM_bias_correction/CAM4_bias_correct.m:
```diff
@@ -0,0 +1,1076 @@
+        ncwriteatt(ncname,'a2x6h_Faxa_swndf','internal_dname','a2x6h')
+        ncwriteatt(ncname,'a2x6h_Faxa_swndf','cell_methods', 'time: mean')
+
+        nccreate(ncname,'a2x6h_Faxa_swndr','Dimensions',{'a2x6h_nx',1,'a2x6h_ny',1,'time',Inf},'Datatype','single','Format','classic');
+        ncwrite(ncname,'a2x6h_Faxa_swndr',swndr_assign)
+        ncwriteatt(ncname,'a2x6h_Faxa_swndr','_FillValue', single(1.e30))
+        ncwriteatt(ncname,'a2x6h_Faxa_swndr','units','W m-2')
+        ncwriteatt(ncname,'a2x6h_Faxa_swndr','long_name','Direct near-infrared incident solar radiation')
+        ncwriteatt(ncname,'a2x6h_Faxa_swndr','standard_name','surface_downward_direct_shortwave_flux_due_to_near_infrared_radiation')
+        ncwriteatt(ncname,'a2x6h_Faxa_swndr','internal_dname','a2x6h')
+        ncwriteatt(ncname,'a2x6h_Faxa_swndr','cell_methods', 'time: mean') 
```
These blocks to write each variable are quite repetitive. You could create a function to create the variable and write its attributes.  That would make your life easier, for example, if you wanted to change the netcdf format to 64-bit Offset you would only need to change the 'format' in one place.

braczka replied:
Thanks @kdraeder and @hkershaw-brown, I added several commits based on your feedback, reducing repetition by generating `load_towermet.m`, `load_CAM.m`, `write_netcdf_att.m` and `plot_diagnostic` functions.   Same functionality as before.  A step in the right direction.

braczka commented:
I pulled off the draft status so it can be reviewed again.   I think I addressed most comments in my updates, but please let me know if things still need to be addressed or better streamlined.   No rush on merging this into main branch of DART.  I am still actively working with site level land DA work which uses this ensemble met forcing as a driver.   The code itself is fairly well commented, and I plan to forgo formal DART webpage documentation at this time, until I kick the wheels a bit more on this.

kdraeder commented on models/clm/CAM_bias_correction/CAM6_bias_correct.m:
```diff
@@ -0,0 +1,650 @@
+
+      end %% Load CAM ensemble loop
+     
+  %% Initialize the ensemble mean scale values to the default of '0' which means no change
```
This is still within the main loop, so it would be helpful to use consistent indenting to show that.

kdraeder commented on models/clm/CAM_bias_correction/CAM6_bias_correct.m:
```diff
@@ -0,0 +1,650 @@
+
+      end
+
+  Faxa_rain_mean= mean(Faxa_rain,1);
```
More indenting.

kdraeder commented on models/clm/CAM_bias_correction/CAM6_bias_correct.m:
```diff
@@ -0,0 +1,650 @@
+      end 
+
+
+  % Assign the adjusted CAM variables to netcdf files for each ensemble member/ year
```
Indenting \#3

kdraeder commented on models/clm/CAM_bias_correction/CAM6_bias_correct.m:
```diff
@@ -0,0 +1,650 @@
+
+      nccreate(ncname,'time','Dimensions',{'time',Inf},'Format','classic');
+      ncwrite(ncname,'time',time_SOLAR)
+          if str2num(yearstr{ii})==2011
```
Indenting \#4.

kdraeder commented on models/clm/CAM_bias_correction/CAM6_bias_correct.m:
```diff
@@ -0,0 +1,650 @@
+      netcdf.close(ncid);
+      nccreate(ncname,'time','Dimensions',{'time',Inf},'Format','classic');
+      ncwrite(ncname,'time',time_1hr)
+          if str2num(yearstr{ii})==2011
```
Indenting \#5.

kdraeder commented on models/clm/CAM_bias_correction/CAM4_bias_correct.m:
```diff
@@ -0,0 +1,511 @@
+    enstr{i}=sprintf('%04d', ens_range(i));
+end
+
+% Check for existence of output file from CAM6_site_grid.sh
```
Should probably be CAM4_site_grid.sh

hkershaw-brown replied:
corrected: 1846d1a35de0137693b6299dfc47c694366b84f3

kdraeder commented on models/clm/CAM_bias_correction/CAM4_bias_correct.m:
```diff
@@ -0,0 +1,511 @@
+% Input data files
+path_CAM = '/glade/work/bmraczka/CAM4_NR1/';
+path_towermet = '/glade/work/bmraczka/SIFMIP2/tower_met_forcing/';
+ens_mem=80;  % CAM6 reanalysis provides 80 total members, 1-80 is valid
```
CAM6->CAM4

kdraeder commented on models/clm/CAM_bias_correction/CAM4_bias_correct.m:
```diff
@@ -0,0 +1,511 @@
+    % CAM4 is 'noleap' calendar, thus trim down tower met to 365 days to match
+    % Remove Feb 29th julian day: 60
+
+      if (str2num(yearstr{ii})==2000 | str2num(yearstr{ii})==2004 | str2num(yearstr{ii})==2008)  
```
Indenting; likely the same as CAM6_bias_correct.m, so I won't flag them all here.

kdraeder commented on models/clm/CAM_bias_correction/CAM4_site_grid.sh:
```diff
@@ -0,0 +1,52 @@
+
+# The US-NR1 flux tower site is used as an example 
+# NR1 location (40.03, -105.55) or (40.03, 254.45)
+# CAM4 grid is 1.95x2.5  or 288 longitude grids and 96 latitude grids
```
Should be "CAM4 grid is 1.95x2.5 degrees, or 96 latitude and 144 longitude points"

hkershaw-brown replied:
grids -> grid points in 351a324e4eaf9c40c958889624097f491c9f4b44

Leaving the 'used for other profiles for a future pull reqeust.

kdraeder commented on models/clm/CAM_bias_correction/CAM4_site_grid.sh:
```diff
@@ -0,0 +1,52 @@
+# This lat/lon corresponds with doma_lon= 102 (255 degrees east) ; doma_lat= 69  ; (40.7368 degrees)
+
+
+for YEAR in {1997..2010..1}
```
I don't use sh much.  Is this indenting standard or helpful?
I can understand indenting the 'for' and 'do' lines the same,
but the contents would be clearer if they were indented more.

kdraeder commented on models/clm/CAM_bias_correction/CAM6_site_grid.sh:
```diff
@@ -0,0 +1,67 @@
+# The US-NR1 flux tower site is used as an example 
+# NR1 location (40.03, -105.55) or (40.03, 254.45)
+# CAM6 grid is 1.25x0.95  or 288 longitude grids and 192 latitude grids
+# This lat/lon corresponds with doma_lon= 204 (255 degrees east) ; doma_lat= 138  ; (40.0524 degrees)
```
Defining the lat and lon points here is even more useful than in CAM4_site_grid.sh.

kdraeder commented on models/clm/CAM_bias_correction/CAM6_site_grid.sh:
```diff
@@ -0,0 +1,67 @@
+for YEAR in {2011..2020..1}
+do
+
+echo "  "
```
Indenting, similar to CAM4_site_grid.sh

kdraeder commented on models/clm/CAM_bias_correction/plot_diagnostic.m:
```diff
@@ -0,0 +1,641 @@
+close all
+
+
+     % For plotting purposes only, make these 3 hour products into 1 hour products for easier comparison 
```
The rest of the script seems to have a lot of redundancy, which could (hopefully) benefit from defining functions.   There might be opportunities above here too,
but I haven't looked in detail at how that would be done.

kdraeder commented on models/clm/CAM_bias_correction/write_netcdf_att.m:
```diff
@@ -0,0 +1,21 @@
+function [] = write_netcdf_att(ncname,varname,fillval,units, ...
+                               long,standard,internal,cell)
+% function write_netcdf: loads all years of tower met forcing (PLUMBER2 format)
```
write_netcdf -> write_netcdf_att
and 'units' is missing.

kdraeder commented on models/clm/CAM_bias_correction/load_towermet.m:
```diff
@@ -0,0 +1,14 @@
+function [tower_main] = load_towermet(metvar,years,path)
+% function load_towermet: loads all years of tower met forcing (PLUMBER2 format)
+% Input:  'metvar' is PLUMBER2 met data name; 'yeartower' is array of tower years to load
```
yeartower -> years

kdraeder commented on models/clm/CAM_bias_correction/load_CAM.m:
```diff
@@ -0,0 +1,22 @@
+% Input:  'varname' is the CAM met name; 'ens' is the ensemble member    
+%         'year' is year of CAM reanalysis
+%         'path' is path to the CAM reanalysis; 'n' indicates Solar (1hr inst),1hr,or 3hr
+% Output: 'cam_main' is the CAM array indexed by ensemble and time
```
...indexed by member?

hkershaw-brown commented:
@braczka  This is is looking much better. Do you mind I go ahead and tidy up the indenting? Just to clear up the comments, see what  is left from Kevin (any dealbreakers).
Otherwise I think this is good to go.

braczka commented:
@hkershaw-brown That would be great -- I have been swamped to this week, so feel free to correct indenting, check on Kevin's comments etc.   Very much appreciate your feedback @kdraeder

hkershaw-brown reviewed (approved):
Approved. 
@braczka I need to merge in this fix https://github.com/NCAR/DART/pull/472
before we can merge your pull request. So I am going go ahead and bundle #472 #468 and this request together as one release.
